sanitize_shenvname returns string with [0-9a-zA-Z_] characters

szemek opened the pull request:
In docker logs for my application I'm getting a lot of noise like this:

```
/etc/profile.d/container_environment.sh: line 10: export: `LIVE.APPLICATION.DEV_PORT_22_TCP_PROTO=tcp': not a valid identifier
/etc/profile.d/container_environment.sh: line 14: export: `LIVE.APPLICATION.DEV_PORT_443_TCP_ADDR=172.17.0.10': not a valid identifier
/etc/profile.d/container_environment.sh: line 24: export: `LIVE.UXPIN.DEV_ENV_MODE=compose': not a valid identifier
```
because it's auto-generated from `docker-compose.yml` configuration

```
app:
    links:
        - balancer:live.application.dev
```

What do you think about restricting characters in env variable name only to the following set `[0-9a-zA-Z_]`  ?

hyperknot commented:
While changing the regex is debatable, moving the code like you did introduces a performance penalty, as the regex is compiled every time it's called.

szemek commented:
@hyperknot you're right. I moved regex pattern back to constant.

hyperknot commented:
Did you test it with your problematic examples?

szemek commented:
@hyperknot yes, I tested problematic cases

Image with warnings:
<img width="1389" alt="with_warnings" src="https://user-images.githubusercontent.com/373530/27503306-47b540ce-587b-11e7-9a06-7ee07083613d.png">

Docker image with updated /etc/my_init
I added to my Dockerfile (with `sed` I replaced regex pattern)
```
RUN sed -i '/SHENV_NAME_WHITELIST_REGEX =/s/.*/SHENV_NAME_WHITELIST_REGEX = re.compile("\\W")/' /sbin/my_init
```

As you can see there are no warnings with exports:
<img width="1389" alt="without_warnings" src="https://user-images.githubusercontent.com/373530/27503342-85fbe6da-587b-11e7-90f7-d17dfe87e1fe.png">

szemek commented:
@hyperknot did you have a chance to take a look at screenshots? :)

hyperknot commented:
sorry for not replying earlier, it looks good but I'll have a proper look
in a few hours

On 2017. Jul 7., Fri at 13:32, Przemysław Dąbek <notifications@github.com>
wrote:

> @hyperknot <https://github.com/hyperknot> did you have a chance to take a
> look at screenshots? :)
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/phusion/baseimage-docker/pull/423#issuecomment-313657445>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAeKjyOh19sSDu-lFYq1AIULmDC4kverks5sLhcxgaJpZM4ODkPV>
> .
>

szemek commented:
@hyperknot sorry for bothering you. Any updates on that? :)

hyperknot commented:
Ah, sorry for not checking it, I'll try to do it today.

On 2017. Jul 17., Mon at 22:01, Przemysław Dąbek <notifications@github.com>
wrote:

> @hyperknot <https://github.com/hyperknot> sorry for bothering you. Any
> updates on that? :)
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/phusion/baseimage-docker/pull/423#issuecomment-315865978>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAeKj73cbaFz8by_pME2h2W6IxleZFOpks5sO72DgaJpZM4ODkPV>
> .
>

hyperknot commented:
OK, just checked the reference about what characters are allowed in environment variables and found it in this SO answer: https://stackoverflow.com/a/2821183/518169

> Environment variable names used by the utilities in the Shell and Utilities volume of IEEE Std 1003.1-2001 consist solely of uppercase letters, digits, and the '_' (underscore) from the characters defined in Portable Character Set and do not begin with a digit.

So `\W` is good enough for us, checking for the "do not begin with a digit" part would be overly pedantic.

szemek commented:
Thanks :)
